### PR TITLE
Fix romanization accuracy

### PR DESCRIPTION
**Pull request by rbelem**
* :wrench: fix(korean): overhaul jong+cho context-dependent assimilation
* :wrench: fix(cyrillic): correct Ukrainian, Serbian, Belarusian mappings
* :sparkles: feat(hindi/punjabi): add inherent vowel handling
* :bug: fix(chinese): preserve tone marks and exclude Japanese from detection
* :sparkles: feat(japanese): use Kuromoji for morphological analysis
* :white_check_mark: test: add comprehensive unit tests for romanization
* :white_check_mark: test: add comprehensive Android instrumented tests

## Summary by CodeRabbit

* **New Features**
  * Optional improved Japanese tokenized romanization support for more accurate readings.

* **Bug Fixes**
  * Enhanced romanization accuracy and context-aware handling for Japanese, Korean, Hindi, Punjabi, and Cyrillic.
  * Language detection refined to avoid misclassifying similar scripts (e.g., Japanese vs Chinese).

* **Tests**
  * Expanded comprehensive JVM test coverage across languages, routing, and many edge cases.

* **Chores**
  * Test setup migrated to JVM unit-test configuration.

[Review Change Stack](https://app.coderabbit.ai/change-stack/shub39/Rush/pull/257)

<details open><summary><h3>Greptile Summary</h3></summary>

This PR replaces the previous Kuromoji download-based approach with a bundled IPADIC TSV for Japanese, overhauls Korean jong+cho context-dependent assimilation into separate `jong`/`CHO_OVERRIDE` maps, adds inherent-vowel handling for Hindi and Punjabi, and corrects several Cyrillic language mappings (Ukrainian, Serbian, Belarusian).

- **Japanese**: `RomanizationUtils` now bundles a 249 k-entry IPADIC reading dictionary and uses a longest-match tokenizer; `JapaneseReadingProvider` loads it on app startup with a proper `Mutex`-guarded idempotent initializer.
- **Korean**: The old combined jong-context map is split into a residual-coda `jong` map and a separate `CHO_OVERRIDE` onset map, enabling correct independent look-up for each resyllabification position.
- **Hindi/Punjabi**: Inherent-vowel `\"a\"` insertion now fires after multi-char conjunct matches as well as single-char consonants, fixing clusters like `kshamaa` and `shakal`.
</details>

<details><summary><h3>Confidence Score: 3/5</h3></summary>

Safe to merge for Japanese, Korean, Hindi, and Punjabi paths; the Ukrainian branch silently produces wrong output for words where ї follows a vowel.

The new Ukrainian ї context rule applies glide-suppression after any Cyrillic predecessor including vowels, so words like наїзд produce naizd instead of nayizd. Two CHO_OVERRIDE entries for aspirated coda consonants (ᆴᄋ and ᆵᄋ) flagged in prior rounds remain uncorrected, producing wrong romanizations for words like 핫아 and 읽어.

shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt — romanizeUkrainianInternal (lines 1074–1082) and CHO_OVERRIDE entries for ᆴᄋ and ᆵᄋ (lines 367–368).
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt | Large overhaul of Korean, Hindi/Punjabi, Japanese, and Cyrillic romanization. The new Ukrainian ї context rule is too broad (matches vowel predecessors). Two CHO_OVERRIDE entries for aspirated coda consonants remain incorrect from a prior round. |
| app/src/main/java/com/shub39/rush/app/JapaneseReadingProvider.kt | New singleton provider with Mutex-guarded initialization and proper reload via resetReadingDictionary(). Looks correct. |
| shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt | Comprehensive instrumented tests with @BeforeClass dictionary loading; assertions relaxed to accept both dictionary-present and absent outcomes. |
| shared/romanization/src/test/kotlin/com/shub39/romanization/RomanizationUtilsIcu4jTest.kt | New JVM unit-test suite for ICU-only romanization paths without dictionary. |
| shared/romanization/src/main/assets/ja_readings.tsv | Bundled IPADIC TSV (~249k entries) for surface-to-katakana reading lookup. |
| app/src/main/java/com/shub39/rush/app/RushApplication.kt | Minimal change: calls JapaneseReadingProvider.initialize on startup instead of the old Kuromoji provider. |

</details>

</details>

<details><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["romanize(text, enabledLanguages)"] --> B{"isJapanese? (kana present)"}
    B -- Yes --> C[romanizeJapanese]
    B -- No --> D{"isKorean?"}
    D -- Yes --> E[romanizeKorean]
    D -- No --> F{"Pure CJK and Japanese enabled?"}
    F -- Yes --> C
    F -- No --> G{"isChinese?"}
    G -- Yes --> H[romanizeChinese Han-Latin ICU]
    G -- No --> I{"isHindi?"}
    I -- Yes --> J[romanizeHindi inherent-vowel]
    I -- No --> K{"isPunjabi?"}
    K -- Yes --> L[romanizePunjabi inherent-vowel]
    K -- No --> M[romanizeCyrillic RU/UK/SR/BG/BE/KY/MK]
    C --> C1{"tokenizerIndex loaded?"}
    C1 -- Yes --> C2[romanizeJapaneseWithTokenizer longest-match IPADIC]
    C1 -- No --> C3[romanizeJapaneseFallback char-by-char]
    E --> E1["Decompose Hangul syllables cho+jung+jong"]
    E1 --> E2["jong map residual coda CHO_OVERRIDE onset"]
```
</details>

<details><summary><h3>Comments Outside Diff (5)</h3></summary>

1. `shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt`, line 1193-1236 ([link](https://github.com/shub39/rush/blob/cfb4792c686a3daf894b82a6ff480ec15ecd8ddc/shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt#L1193-L1236)) 

   <a href="#">P1</a> **Missing inherent 'a' after two-char Gurmukhi consonants**

   When a two-char Gurmukhi consonant (ਸ਼, ਖ਼, ਗ਼, ਜ਼, ਫ਼, ਲ਼) is consumed via the `twoChar` lookup, the `consumed = true` path skips the entire inherent-'a' block. If the next character is a consonant, the 'a' is silently dropped.

   Concrete failure: "ਸ਼ਕਲ" (shakal) processes as `sh`+`k`+`a`+`l` → `"shkal"` instead of `"shakal"`. The inherent 'a' that should appear between "sh" and "k" is never appended because the code only enters the `if (!consumed)` branch for single-char matches.

   The same two-char structure in `romanizeHindi` is unaffected in practice (no 2-char Devanagari keys exist in the map), but the logic gap is present in `romanizePunjabi` for all six nukta-combined consonants.

2. `shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt`, line 1144-1153 ([link](https://github.com/shub39/rush/blob/2a85c8301463bf8f2d1ed793d4030807c4546dff/shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt#L1144-L1153)) 

   <a href="#">P1</a> The `romanizeHindi` function only performs a 2-char lookahead, but `DEVANAGARI_ROMAJI_MAP` contains 3-character conjunct keys (`"क्ष"→"ksh"`, `"त्र"→"tr"`, `"ज्ञ"→"gy"`, `"श्र"→"shr"`). These entries are never reached, so `"ज्ञ"` is transliterated character-by-character as `"j"+""+"ny"` = `"jny"` instead of `"gy"`. Common words like `ज्ञान` (gyan) silently produce `"jnyaan"`. A 3-char lookahead — identical to the pattern already used in `romanizeRussianInternal` — needs to run before the 2-char check.

3. `shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt`, line 1234-1243 ([link](https://github.com/shub39/rush/blob/785088c8f2841078091238753bc24192c8567181/shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt#L1234-L1243)) 

   <a href="#">P1</a> **Missing inherent 'a' after two-char Gurmukhi consonant matches in `romanizePunjabi`**

   When a two-char Gurmukhi consonant (`"ਸ਼"`, `"ਖ਼"`, `"ਗ਼"`, `"ਜ਼"`, `"ਫ਼"`, `"ਲ਼"`) is consumed via the `twoChar` lookup (`consumed = true`), the entire `if (!consumed)` block — which appends the inherent 'a' before an unvowelled following consonant — is skipped. Concretely, `"ਸ਼ਕਲ"` (shakal) produces `"shkal"` instead of `"shakal"`: after `"ਸ਼"` is consumed as `"sh"`, no 'a' is inserted before `"ਕ"`. The inherent-'a' check needs to also run when a two-char key is consumed.

4. `shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt`, line 830-841 ([link](https://github.com/shub39/rush/blob/8d10c719133f214d291b94a3a0771309dbc04217/shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt#L830-L841)) 

   <a href="#">P1</a> **`testRomanize_autoDetectChinese` will always fail**

   With both "Japanese" and "Chinese" in `enabledLanguages`, the `romanize()` dispatch at lines 1429–1431 matches any pure-CJK text before the Chinese branch can fire: `text.any { it in '\u4E00'..'\u9FFF' }` is true for "你好", so the call routes to `romanizeJapanese`. Since "你好" is absent from the IPADIC dictionary, the tokenizer passes both characters through unchanged, producing "你 好". Neither `contains("nǐ")` nor `contains("ni")` will be true, so the assertion fails on every instrumented run.

   The sibling test `testChinese_pureCjkWithBothJapaneseAndChinese` already documents the correct expected behaviour with `assertTrue(result!!.isNotEmpty())`. This test should use the same weak assertion (or be moved to a Chinese-only `enabledLanguages` list) to match the documented intent.

5. `shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt`, line 1205-1216 ([link](https://github.com/shub39/rush/blob/9f7a00eb732cd56d1c18e4d7d6058d2701b7ba85/shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt#L1205-L1216)) 

   <a href="#">P1</a> **`testRomanize_autoDetectChinese` assertion always fails at runtime**

   When both "Japanese" and "Chinese" are in `enabledLanguages`, `romanize("你好", ...)` matches the pure-CJK branch first and calls `romanizeJapanese`. The tokenizer processes '你' and '好' as two separate single-character tokens and joins them with `joinToString(" ")`, producing `"你 好"`. The assertion `result!!.contains("你好")` evaluates to false, and `result!!.contains("nǐ")` is also false. The sibling test `testChinese_pureCjkWithBothJapaneseAndChinese` already uses the correct weak assertion `assertTrue(result!!.isNotEmpty())`.
</details>

<sub>Reviews (50): Last reviewed commit: [":lock: fix(romanization): add Mutex guar..."](https://github.com/shub39/rush/commit/9a084c4a4edc0300f3bbe5a7cc7d59ca51bc1f40) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=31847720)</sub>

> Greptile also left **1 inline comment** on this PR.

### CODE REVIEW

**Comment by rbelem** [comment]
@coderabbitai review

File: shared/romanization/build.gradle.kts [hunk 1]
```diff
@@ -45,10 +45,11 @@ android {
 
 dependencies {
     implementation(libs.kotlinx.coroutines)
+    implementation("com.atilika.kuromoji:kuromoji-ipadic:0.9.0")
```
[the remark this answers is not in the document]
Comment (rbelem) [reply shared/romanization/build.gradle.kts]: Fixed!

File: shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt [hunk 2]
```diff
@@ -22,58 +22,405 @@ import org.junit.Test
...
+    fun testHindi_hindi() = runTest {
+        val result = RomanizationUtils.romanizeHindi("हिंदी")
+        assertEquals("hindi", result)
+    }
+
+    @Test
+    fun testHindi_pyar() = runTest {
+        val result = RomanizationUtils.romanizeHindi("प्यार")
+        assertEquals("pyar", result)
+    }
```
[the remark this answers is not in the document]
Comment (rbelem) [reply shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt]: Fixed!

File: shared/romanization/src/test/kotlin/com/shub39/romanization/RomanizationUtilsIcu4jTest.kt [hunk 3]
```diff
@@ -0,0 +1,894 @@
...
+    }
+
+    @Test
+    fun testKorean_doubleConsonantBatchim() = runTest {
+        // Known issue: context-dependent jong mapping produces extra consonant
+        // 삶은 → "salgeun" (not "salgeun" as expected from batchim rules)
+        val result = RomanizationUtils.romanizeKorean("삶은")
+        assertTrue(result.contains("sal"))
+        assertTrue(result.contains("eun"))
+    }
```
[the remark this answers is not in the document]
Comment (rbelem) [reply shared/romanization/src/test/kotlin/com/shub39/romanization/RomanizationUtilsIcu4jTest.kt]: Fixed!

**Comment by rbelem** [comment]
@coderabbitai review

File: shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt [hunk 4]
```diff
@@ -748,6 +939,54 @@ object RomanizationUtils {
...
+            "ष",
+            "स",
+            "ह",
+            "ळ",
+            "क्ष",
+            "त्र",
+            "ज्ञ",
+            "श्र",
+            "़",
+        )
```
[the remark this answers is not in the document]
Comment (rbelem) [reply shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt]: Fixed!

**Comment by shub39** [comment]
release apk size still remains larger than 10mb

**Comment by rbelem** [comment]
> release apk size still remains larger than 10mb

I will look for an alternative.

**Comment by rbelem** [comment]
@shub39 would downloading the dictionary be a good option?

**Comment by shub39** [comment]
> @shub39 would downloading the dictionary be a good option?

yes that would be better

**Comment by rbelem** [comment]
@shub39 it should be better now.

**Comment by shub39** [comment]
@rbelem could you explain what's the python script for in romanization assets? 
also how it works? is `ja_readings.tsv` replacing the need to download assets?

**Comment by rbelem** [comment]
@shub39 The goal is to replace the libraries and create a more lightweight dictionary. Should it be committed to the repository or generated using the build pipeline?

**Comment by shub39** [comment]
it'll be better to have it in repository. so with this the user do not need to download dictionaries after installation?

File: app/src/main/java/com/shub39/rush/app/JapaneseReadingProvider.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            return waitForInitialization()
+        }
+
+        isInitializing = true
+        try {
+            return doInitialize(context)
+        } finally {
+            isInitializing = false
+        }
+    }
```
[the remark this answers is not in the document]
Comment (rbelem) [reply app/src/main/java/com/shub39/rush/app/JapaneseReadingProvider.kt]: Fixed

File: shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt [hunk 6]
```diff
@@ -14,66 +14,510 @@
...
+        assertTrue(result.contains("ksh"))
+        assertTrue(result.contains("maa"))
+    }
+
+    @Test
+    fun testHindi_conjunctGy() = runTest {
+        val result = RomanizationUtils.romanizeHindi("ज्ञान")
+        assertNotNull(result)
         assertTrue(result.isNotEmpty())
     }
```
[the remark this answers is not in the document]
Comment (rbelem) [reply shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt]: Fixed

File: shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt [hunk 7]
```diff
@@ -748,31 +1092,141 @@ object RomanizationUtils {
...
+                                    nextChar == "ँ")
+                        ) {
+                            sb.append("a")
+                        }
+                    }
                     i++
                 }
             }
             sb.toString()
         }
```
[the remark this answers is not in the document]
Comment (rbelem) [reply shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt:1296]: fixed

File: shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt [hunk 8]
```diff
@@ -14,66 +14,509 @@
...
+        val result = RomanizationUtils.romanizeJapanese("hello")
+        assertEquals("hello", result)
+    }
+
+    @Test
+    fun testJapanese_katakanaLove() = runTest {
+        val result = RomanizationUtils.romanizeJapanese("アイシテル")
+        assertNotNull(result)
         assertTrue(result.isNotEmpty())
     }
```
[the remark this answers is not in the document]
Comment (rbelem) [reply shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt]: Fixed

File: shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt [hunk 9]
```diff
@@ -905,8 +1422,15 @@ object RomanizationUtils {
...
+            // Japanese text with kana → always treat as Japanese
             "Japanese" in enabledLanguages && isJapanese(text) -> romanizeJapanese(text)
             "Korean" in enabledLanguages && isKorean(text) -> romanizeKorean(text)
+            // Pure CJK (no kana): route to Japanese first. The IPADIC reading
+            // dictionary covers common Japanese kanji words. Chinese text that
+            // is not in the dictionary falls through to romanizeChinese below.
+            "Japanese" in enabledLanguages &&
+                text.any { it in '\u4E00'..'\u9FFF' || it in '\u3400'..'\u4DBF' } ->
+                romanizeJapanese(text)
             "Chinese" in enabledLanguages && isChinese(text) -> romanizeChinese(text)
```
[the remark this answers is not in the document]
Comment (rbelem) [reply shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt:1515]: Fixed

**Comment by rbelem** [comment]
@shub39 yes, it will now be included in the apk.

File: shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt [hunk 10]
```diff
@@ -748,31 +1101,160 @@ object RomanizationUtils {
...
+                                        nextChar !in DEVANAGARI_REPLACING_VOWEL_SIGNS &&
+                                        nextChar != DEVANAGARI_VIRAMA &&
+                                        (nextChar in DEVANAGARI_CONSONANTS ||
+                                            nextChar == "ं" ||
+                                            nextChar == "ः" ||
+                                            nextChar == "ँ")
+                                ) {
+                                    sb.append("a")
+                                }
+                            }
```
[the remark this answers is not in the document]
Comment (rbelem) [reply shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt]: Fixed

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

File: shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt [hunk 11]
```diff
@@ -79,153 +146,264 @@ object RomanizationUtils {
...
+            "ᆺᄆ" to "m",
+            "ᆻᄆ" to "m",
+            "ᆽᄆ" to "m",
+            "ᆾᄆ" to "m",
+            "ᆿᄆ" to "m",
+            "ᇀᄆ" to "m",
+            "ᇂᄆ" to "m",
+            "ᆬᄂ" to "n",
+            "ᆶᄆ" to "m",
+        )
```
[the remark this answers is not in the document]
Comment (rbelem) [reply shared/romanization/src/main/kotlin/com/shub39/romanization/RomanizationUtils.kt:477]: Fixed!

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

File: shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt [hunk 12]
```diff
@@ -14,66 +14,541 @@
...
+        val result = RomanizationUtils.romanizeJapanese("へや")
+        assertTrue(result.contains("heya"))
+    }
+
+    @Test
+    fun testJapanese_particleO() = runTest {
+        val result = RomanizationUtils.romanizeJapanese("それを")
+        // Without morphological analyzer, を is romanized as "wo"
+        assertEquals("sorewo", result)
+    }
```
[the remark this answers is not in the document]
Comment (rbelem) [reply shared/romanization/src/androidTest/kotlin/RomanizationUtilsTest.kt]: Fixed

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@shub39 i think it is ready to be merged.

**Comment by rbelem** [comment]
@greptile review

**Comment by shub39** [comment]
since the dictionary is part of the codebase the python script can be removed then? also I see there's context in the constructor parameter of LyricsVM? is that necessary? shouldn't all the changes be isolated to the romanization module?

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@shub39 I updated romanization to be isolated.

**Comment by rbelem** [comment]
@greptile review

**Comment by rbelem** [comment]
@shub39 I would prefer to keep the python script to generate new dictionary versions in the future.

**Comment by rbelem** [comment]
I apologize for using up all the Greptile credits. 😞

**Comment by shub39** [comment]
no the script should be removed, i don't think the language will change in future to need dictionary regeneration. also the `JapaneseReadingProvider.kt` file is not needed. There should not be any changes in the app module

**Comment by rbelem** [comment]
> no the script should be removed, i don't think the language will change in future to need dictionary regeneration. also the `JapaneseReadingProvider.kt` file is not needed. There should not be any changes in the app module

Python script and dead code removed.

**Comment by rbelem** [comment]
@shub39 Is there anything you'd like me to do with this pull request??

**Review by shub39 — APPROVED** [review]
lgtm
